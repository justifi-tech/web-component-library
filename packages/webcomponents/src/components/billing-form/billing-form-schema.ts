import { object, string } from 'yup';
import { RegExZip } from '../../utils/utils';

// All of these schema validations already exist in the schema files in business-form directory. 
// I want to make a common schema directory for top level access to all schema files. I guess?

export interface BillingFormFields {
  name: string;
  address_line1: string;
  address_line2?: string;
  address_city: string;
  address_state: string;
  address_postal_code: string;
}

const BillingFormSchema = object({
  name: string().required('Enter name'),
  address_line1: string().required('Enter street address'),
  address_line2: string(),
  address_city: string().required('Enter city'),
  address_state: string().required('Choose state'),
  address_postal_code: string().required('Enter ZIP').matches(RegExZip, 'Enter a valid ZIP').min(5, 'Enter a valid ZIP'),
});

export default BillingFormSchema;
