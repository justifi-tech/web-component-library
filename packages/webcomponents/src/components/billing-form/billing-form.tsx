import { Component, Host, h, State, Listen, Method, Prop } from '@stencil/core';
import { ValidationError } from 'yup';
import BillingFormSchema, { BillingFormFields } from './billing-form-schema';
import StateOptions from './state-options';

/**
 * @exportedPart label: Label for inputs
 * @exportedPart input: The input fields
 * @exportedPart input-invalid: Invalid state for inputs
 */
@Component({
  tag: 'justifi-billing-form',
  styleUrl: 'billing-form.scss',
  shadow: true,
})
export class BillingForm {
  /**
   * (Optional) A label for the form.
   */
  @Prop({ mutable: true }) legend?: string;
  @State() billingFields: BillingFormFields = {
    name: '',
    address_line1: '',
    address_line2: '',
    address_city: '',
    address_state: '',
    address_postal_code: '',
  };

  @State() billingFieldsErrors: any = {};

  @Listen('fieldReceivedInput')
  setFormValue(event) {
    const data = event.detail;
    const billingFieldsClone = { ...this.billingFields };
    if (data.name) {
      billingFieldsClone[data.name] = data.value;
      this.billingFields = billingFieldsClone;
    }
  }

  /**
   * Method for filling the form with provided data
   * @argument {BillingFormFields} fields - The fields to fill the form with
   */
  @Method()
  async fill(fields: BillingFormFields) {
    this.billingFields = { ...fields };
  }

  /**
   * Run validation on the form
   */
  @Method()
  async validate() {
    const newErrors = {};
    let isValid: boolean = true;

    try {
      await BillingFormSchema.validate(this.billingFields, { abortEarly: false });
    } catch (err) {
      isValid = false;
      err.inner.map((item: ValidationError) => {
        newErrors[item.path] = item.message;
      });
    }

    this.billingFieldsErrors = newErrors;

    return { isValid: isValid };
  }

  /**
   * Returns the values of the form as an object
   * @returns {Promise<BillingFormFields>} The values of the form
   */
  @Method()
  async getValues(): Promise<BillingFormFields> {
    return this.billingFields;
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <fieldset>
          {this.legend && <legend>{this.legend}</legend>}
          <div class="row gy-3">
            <div class="col-12">
              <text-input name="name" label="Full Name" defaultValue={this.billingFields.name} error={this.billingFieldsErrors.name} />
            </div>
            <div class="col-12">
              <text-input name="address_line1" label="Street Address" defaultValue={this.billingFields.address_line1} error={this.billingFieldsErrors.address_line1} />
            </div>
            <div class="col-12">
              <text-input
                name="address_line2"
                label="Apartment, Suite, etc. (optional)"
                defaultValue={this.billingFields.address_line2}
                error={this.billingFieldsErrors.address_line2}
              />
            </div>
            <div class="col-12">
              <text-input name="address_city" label="City" defaultValue={this.billingFields.address_city} error={this.billingFieldsErrors.address_city} />
            </div>
            <div class="col-12">
              <select-input
                name="address_state"
                label="State"
                options={StateOptions}
                defaultValue={this.billingFields.address_state}
                error={this.billingFieldsErrors.address_state}
              />
            </div>
            <div class="col-12">
              <text-input name="address_postal_code" label="ZIP" defaultValue={this.billingFields.address_postal_code} error={this.billingFieldsErrors.address_postal_code} />
            </div>
          </div>
        </fieldset>
      </Host>
    );
  }
}
