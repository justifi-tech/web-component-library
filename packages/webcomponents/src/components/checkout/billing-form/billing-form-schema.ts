import { object } from 'yup';
import {
  cityValidation,
  identityNameValidation,
  lineOneValidation,
  lineTwoValidation,
  postalValidation,
  stateValidation,
} from '../../business-forms/schemas/schema-validations';

export interface BillingFormFields {
  name: string;
  address_line1: string;
  address_line2?: string;
  address_city: string;
  address_state: string;
  address_postal_code: string;
}

export const fullBillingSchema = () => object({
  name: identityNameValidation.required('Enter full name'),
  address_line1: lineOneValidation.required('Enter street address'),
  address_line2: lineTwoValidation.nullable(),
  address_city: cityValidation.required('Enter city'),
  address_state: stateValidation.required('Select state'),
  address_postal_code: postalValidation.required('Enter postal code')
});

export const postalOnlySchema = () => object({
  name: identityNameValidation.nullable(),
  address_line1: lineOneValidation.nullable(),
  address_line2: lineTwoValidation.nullable(),
  address_city: cityValidation.nullable(),
  address_state: stateValidation.nullable(),
  address_postal_code: postalValidation.required('Enter postal code')
});

export const emptyBillingSchema = () => object({
  name: identityNameValidation.nullable(),
  address_line1: lineOneValidation.nullable(),
  address_line2: lineTwoValidation.nullable(),
  address_city: cityValidation.nullable(),
  address_state: stateValidation.nullable(),
  address_postal_code: postalValidation.nullable()
});
