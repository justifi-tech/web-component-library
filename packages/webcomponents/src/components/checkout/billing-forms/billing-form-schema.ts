import { object } from 'yup';
import {
  cityValidation,
  identityNameValidation,
  lineOneValidation,
  lineTwoValidation,
  postalValidation,
  stateValidation,
} from '../../business-forms/schemas/schema-validations';

// BillingFormFields and BillingFormSchema are used for justifi-billing-form which renders the entire billing form.
export interface BillingFormFields {
  name: string;
  address_line1: string;
  address_line2?: string;
  address_city: string;
  address_state: string;
  address_postal_code: string;
}

export const BillingFormSchema = object({
  name: identityNameValidation.required('Enter full name'),
  address_line1: lineOneValidation.required('Enter street address'),
  address_line2: lineTwoValidation.nullable(),
  address_city: cityValidation.required('Enter city'),
  address_state: stateValidation.required('Select state'),
  address_postal_code: postalValidation.required('Enter postal code'),
});

// PostalFormFields and PostalFormSchema are used for justifi-postal-form which renders the postal code field only.
export interface PostalFormFields {
  address_postal_code: string;
}

export const PostalFormSchema = object({
  address_postal_code: postalValidation.required('Enter postal code'),
});

export function isBillingFormFields(fields: any): fields is BillingFormFields {
  return (
    'name' in fields &&
    'address_line1' in fields &&
    'address_city' in fields &&
    'address_state' in fields &&
    'address_postal_code' in fields
  );
}

export function isPostalFormFields(fields: any): fields is PostalFormFields {
  return (
    typeof fields === 'object' &&
    fields !== null &&
    Object.keys(fields).length === 1 &&
    'address_postal_code' in fields
  );
}
