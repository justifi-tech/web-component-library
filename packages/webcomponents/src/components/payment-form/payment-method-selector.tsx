import { Component, Event, EventEmitter, Fragment, h, Prop } from '@stencil/core';
import { PaymentMethodTypes } from '../../api';

const PaymentMethodLabels = {
  bankAccount: 'Bank Account',
  card: 'Card',
};

@Component({
  tag: 'justifi-payment-method-selector',
  styleUrl: 'payment-method-selector.css',
})
export class PaymentMethodSelector {
  @Prop() selectedPaymentMethodType: PaymentMethodTypes;
  @Event({ bubbles: true }) paymentMethodSelected: EventEmitter;

  paymentMethodTypes: PaymentMethodTypes[] = [PaymentMethodTypes.card, PaymentMethodTypes.bankAccount];
  defaultRadioButtonOption!: HTMLInputElement;

  onChangeHandler(event: any) {
    this.paymentMethodSelected.emit(event.target.value);
  }

  render() {
    return (
      <div class="btn-group jfi-btn-radio-group" role="group" aria-label="Radio toggle group for payment method">
        {this.paymentMethodTypes.map((paymentMethodType: PaymentMethodTypes) => (
          <Fragment>
            <input
              id={paymentMethodType}
              type="radio"
              name="paymentMethodType"
              value={paymentMethodType}
              onChange={(event: any) => this.onChangeHandler(event)}
              checked={this.selectedPaymentMethodType === paymentMethodType}
              class="btn-check jfi-btn-radio"
            />
            <label htmlFor={paymentMethodType} class="btn btn-outline-primary jfi-btn-radio-label">
              {PaymentMethodLabels[paymentMethodType]}
            </label>
          </Fragment>
        ))}
      </div>
    );
  }
}
