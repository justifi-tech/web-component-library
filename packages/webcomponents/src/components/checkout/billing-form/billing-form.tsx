import { Component, h, Prop, Method } from '@stencil/core';
import { BillingFormFields } from './billing-form-schema';

@Component({
  tag: 'justifi-billing-form',
})
export class BillingForm {
  @Prop({ mutable: true }) legend?: string;
  @Prop() hideCardBillingForm?: boolean;
  @Prop() hideBankAccountBillingForm?: boolean;
  @Prop() paymentMethodType?: string;

  private selectedFormRef?: any;

  private get showSimpleCardBillingForm() {
    return this.paymentMethodType === 'card' && this.hideCardBillingForm;
  }

  private get showSimpleBankAccountBillingForm() {
    return this.paymentMethodType === 'bankAccount' && this.hideBankAccountBillingForm;
  }

  @Method()
  async getValues(): Promise<BillingFormFields> {
    return this.selectedFormRef?.getValues();
  }

  @Method()
  async fill(fields: BillingFormFields) {
    this.selectedFormRef?.fill(fields);
  }

  @Method()
  async validate(): Promise<{ isValid: boolean, errors: any }> {
    return this.selectedFormRef?.validate();
  }

  render() {
    if (this.showSimpleBankAccountBillingForm) {
      return (
        <justifi-bank-account-billing-form-simple
          legend={this.legend}
          ref={(el: any) => (this.selectedFormRef = el)}
        />
      );
    }

    if (this.showSimpleCardBillingForm) {
      return (
        <justifi-card-billing-form-simple
          legend={this.legend}
          ref={(el: any) => (this.selectedFormRef = el)}
        />
      );
    }

    return (
      <justifi-billing-form-full
        legend={this.legend}
        ref={(el) => (this.selectedFormRef = el)}
      />
    );
  }
}
