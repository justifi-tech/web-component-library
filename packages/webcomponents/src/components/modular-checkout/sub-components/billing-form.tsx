import { Component, h, State, Method } from '@stencil/core';
import { BillingFormFields, billingFormSchema } from './billing-form-schema';
import { billingForm } from '../../styles/parts';
import { numberOnlyHandler } from '../../ui-components/form/utils';
import StateOptions from '../../utils/state-options';
import { FormController } from '../../ui-components/form/form';
import { StyledHost } from '../../ui-components';

@Component({
  tag: 'justifi-billing-information-form',
  shadow: true
})
export class BillingInformationForm {
  @State() formController: FormController;
  @State() billingInfo: {}
  @State() errors: any = {};

  componentWillLoad() {
    this.formController = new FormController(billingFormSchema());
  }

  componentDidLoad() {
    this.formController.values.subscribe(values =>
      this.billingInfo = { ...values }
    );
    this.formController.errors.subscribe(errors => {
      this.errors = { ...errors };
    });
  }

  inputHandler = (name: string, value: string) => {
    this.formController.setValues({
      ...this.formController.values.getValue(),
      [name]: value,
    });
  }

  @Method()
  async getValues(): Promise<BillingFormFields> {
    return this.formController.values.getValue();
  }

  @Method()
  async fill(fields: BillingFormFields) {
    this.formController.setInitialValues(fields);
  }

  @Method()
  async validate(): Promise<{ isValid: boolean }> {
    let isValid: boolean = await this.formController.validate();
    return { isValid: isValid };
  }

  render() {

    const billingFormDefaultValue = this.formController.getInitialValues();

    return (
      <StyledHost>
        <div part={billingForm}>
          <form>
            <fieldset>
              <div class="row gy-3">
                <div class="col-12">
                  <form-control-text
                    name='name'
                    label='Full Name'
                    defaultValue={billingFormDefaultValue.name}
                    errorText={this.errors.name}
                    inputHandler={this.inputHandler}
                  />
                </div>
                <div class="col-12">
                  <form-control-text
                    name='address_line1'
                    label='Street Address'
                    defaultValue={billingFormDefaultValue.address_line1}
                    errorText={this.errors.address_line1}
                    inputHandler={this.inputHandler}
                  />
                </div>
                <div class="col-12">
                  <form-control-text
                    name='address_line2'
                    label="Apartment, Suite, etc. (optional)"
                    defaultValue={billingFormDefaultValue.address_line2}
                    errorText={this.errors.address_line2}
                    inputHandler={this.inputHandler}
                  />
                </div>
                <div class="col-12">
                  <form-control-text
                    name='address_city'
                    label="City"
                    defaultValue={billingFormDefaultValue.address_city}
                    errorText={this.errors.address_city}
                    inputHandler={this.inputHandler}
                  />
                </div>
                <div class="col-12">
                  <form-control-select
                    name='address_state'
                    label='State'
                    options={StateOptions}
                    defaultValue={billingFormDefaultValue.address_state}
                    errorText={this.errors.address_state}
                    inputHandler={this.inputHandler}
                  />
                </div>
                <div class="col-12">
                  <form-control-text
                    name='address_postal_code'
                    label="ZIP"
                    defaultValue={billingFormDefaultValue.address_postal_code}
                    errorText={this.errors.address_postal_code}
                    inputHandler={this.inputHandler}
                    maxLength={5}
                    keyDownHandler={numberOnlyHandler}
                  />
                </div>
              </div>
            </fieldset>
          </form>
        </div>
      </StyledHost>
    );
  }
}
