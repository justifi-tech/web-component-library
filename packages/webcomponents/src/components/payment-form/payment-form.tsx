import { Component, Prop, h, Host, State, Method, Event, EventEmitter } from '@stencil/core';
import { PaymentMethodTypes } from '../../api';
import { BillingFormFields } from '../billing-form/billing-form-schema';
import { CreatePaymentMethodResponse } from '../payment-method-form/payment-method-responses';
import { loadFontsOnParent } from '../../utils/utils';
import { config } from '../../../config';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorMessage } from '../../api/services/utils';

@Component({
  tag: 'justifi-payment-form',
  styleUrl: 'payment-form.scss',
  shadow: true,
})
export class PaymentForm {
  @Prop() bankAccount?: boolean;
  @Prop() card?: boolean = true;
  @Prop() email?: string;
  @Prop() clientId?: string;
  @Prop() authToken?: string;
  @Prop() accountId?: string;
  @Prop() submitButtonText?: string;

  @State() submitButtonEnabled: boolean = true;
  @State() isLoading: boolean = false;
  @State() selectedPaymentMethodType: PaymentMethodTypes = PaymentMethodTypes.card;

  @Event() submitted: EventEmitter<CreatePaymentMethodResponse>;
  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  private paymentMethodFormRef?: HTMLJustifiPaymentMethodFormElement;
  private billingFormRef?: HTMLJustifiBillingFormElement;

  componentWillLoad() {
    if (!this.validateProps()) {
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        message: 'clientId or authToken is required',
        severity: ComponentErrorSeverity.ERROR
      });
      this.submitButtonEnabled = false;
    }
  }

  connectedCallback() {
    loadFontsOnParent();
  }

  @Method()
  async fillBillingForm(fields: BillingFormFields) {
    this.billingFormRef.fill(fields);
  }

  @Method()
  async enableSubmitButton() {
    this.submitButtonEnabled = true;
  }

  @Method()
  async disableSubmitButton() {
    this.submitButtonEnabled = false;
  }

  showPaymentMethodTypeSelector() {
    return this.card && this.bankAccount;
  }

  getSelectedPaymentMethodType() {
    if (this.showPaymentMethodTypeSelector()) {
      return this.selectedPaymentMethodType;
    } else if (this.card) {
      return PaymentMethodTypes.card;
    } else if (this.bankAccount) {
      return PaymentMethodTypes.bankAccount;
    } else {
      return PaymentMethodTypes.card;
    }
  }

  paymentMethodSelectedHandler(event: CustomEvent) {
    const paymentMethodType: PaymentMethodTypes = event.detail;
    this.selectedPaymentMethodType = paymentMethodType;
  }

  private validateProps(): boolean {
    return !!(this.clientId || this.authToken);
  }

  private getToken(): string {
    return this.authToken || this.clientId;
  }

  async submit(event) {
    event.preventDefault();
    if (!this.paymentMethodFormRef || !this.billingFormRef) return;

    const billingFormValidation = await this.billingFormRef.validate();
    const paymentMethodFormValidation = await this.paymentMethodFormRef.validate();

    if (!billingFormValidation.isValid || !paymentMethodFormValidation.isValid) return;

    this.isLoading = true;

    try {
      const billingFormFieldValues = await this.billingFormRef.getValues();
      const paymentMethodData = { email: this.email, ...billingFormFieldValues };
      const tokenizeResponse = await this.paymentMethodFormRef.tokenize(this.getToken(), paymentMethodData, this.accountId);
      if (tokenizeResponse.error) {
        this.errorEvent.emit({
          errorCode: ComponentErrorCodes.TOKENIZE_ERROR,
          message: tokenizeResponse.error.message,
          severity: ComponentErrorSeverity.ERROR
        });
      }
      this.submitted.emit(tokenizeResponse);
    } catch (error) {
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.UNKNOWN_ERROR,
        message: getErrorMessage(error),
        severity: ComponentErrorSeverity.ERROR
      })
    } finally {
      this.isLoading = false;
    }
  }

  render() {
    return (
      <Host>
        <form class="row gy-3">
          {this.showPaymentMethodTypeSelector() && (
            <div class="col-12">
              <justifi-payment-method-selector
                selectedPaymentMethodType={this.getSelectedPaymentMethodType()}
                onPaymentMethodSelected={event => this.paymentMethodSelectedHandler(event)}
              />
            </div>
          )}
          <div class="col-12">
            <justifi-payment-method-form
              payment-method-form-type={this.getSelectedPaymentMethodType()}
              iframeOrigin={config.iframeOrigin}
              ref={el => {
                if (el) {
                  this.paymentMethodFormRef = el;
                }
              }}
            />
          </div>
          <div class="col-12">
            <justifi-billing-form
              legend="Billing Info"
              ref={el => {
                if (el) {
                  this.billingFormRef = el;
                }
              }}
            />
          </div>
          <slot name='insurance' />
          <div class="col-12">
            <button
              data-testid="submit-button"
              type="submit"
              onClick={event => this.submit(event)}
              disabled={!this.submitButtonEnabled || this.isLoading}
              class={`btn btn-primary jfi-submit-button${this.isLoading ? ' jfi-submit-button-loading' : ''}`}
            >
              {
                this.isLoading ?
                  <div class="spinner-border spinner-border-sm" role="status">
                    <span class="visually-hidden">Loading...</span>
                  </div> :
                  this.submitButtonText || 'Submit'
              }
            </button>
          </div>
        </form>
      </Host>
    );
  }
}
