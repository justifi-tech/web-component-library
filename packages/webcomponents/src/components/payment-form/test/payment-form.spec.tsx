import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { PaymentForm } from '../payment-form';
import { PaymentMethodTypes } from '../../../api';
import { BillingFormFields } from '../../billing-form/billing-form-schema';
import { PaymentMethodSelector } from '../payment-method-selector';
import { PaymentMethodForm } from '../../payment-method-form/payment-method-form';
import { BillingForm } from '../../billing-form/billing-form';

describe('justifi-payment-form', () => {
  it('should throw error if both clientId and authToken are not provided', async () => {
    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [PaymentForm],
      template: () => <justifi-payment-form onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        errorCode: 'missing-props',
        message: 'clientId or authToken is required',
        severity: 'error',
      }
    }));
  });

  it('should update submitButtonEnabled state when enableSubmitButton method is called', async () => {
    // Mock component
    const mockComponent = new PaymentForm();
    mockComponent.submitButtonEnabled = false;

    // Call the method
    await mockComponent.enableSubmitButton();

    // Assertion
    expect(mockComponent.submitButtonEnabled).toBe(true);
  });

  it('should update selectedPaymentMethodType when paymentMethodSelected event is fired', async () => {
    // Mock component
    const mockComponent = new PaymentForm();
    mockComponent.selectedPaymentMethodType = PaymentMethodTypes.card;

    // Define the event details
    const detail = PaymentMethodTypes.bankAccount;
    const event = new CustomEvent('paymentMethodSelected', { detail });

    // Simulate the event
    mockComponent.paymentMethodSelectedHandler(event);

    // Assertion
    expect(mockComponent.selectedPaymentMethodType).toBe(PaymentMethodTypes.bankAccount);
  });

  // Props Testing
  it('should have the expected default properties', () => {
    const paymentForm = new PaymentForm();

    // Assert that the properties exist and have the expected default values
    expect(paymentForm).toHaveProperty('bankAccount', undefined);
    expect(paymentForm).toHaveProperty('card', true);
    expect(paymentForm).toHaveProperty('email', undefined);
    expect(paymentForm).toHaveProperty('clientId', undefined);
    expect(paymentForm).toHaveProperty('accountId', undefined);
    expect(paymentForm).toHaveProperty('submitButtonText', undefined);
  });

  // Method Testing
  it('should correctly fill billing form when fillBillingForm method is called', async () => {
    // Mock child component
    const mockBillingFormRef = {
      fill: jest.fn(),
    };

    // Mock component
    const mockComponent = new PaymentForm();

    // Cast to any to bypass type checker and directly set private property
    (mockComponent as any).billingFormRef = mockBillingFormRef;

    // Define the fields to fill the billing form
    const fields: BillingFormFields = {
      name: 'John Doe',
      address_line1: '123 Main St',
      address_line2: 'Apt 4B',
      address_city: 'Townsville',
      address_state: 'TS',
      address_postal_code: '12345',
    };

    // Call the method
    await mockComponent.fillBillingForm(fields);

    // Check that the child component's fill method was called with the correct arguments
    expect(mockBillingFormRef.fill).toHaveBeenCalledWith(fields);
  });

  it('should submit the form correctly and trigger the "submitted" event with the correct payload when submit method is called', async () => {
    // Mock billingFormRef and paymentMethodFormRef
    const mockBillingFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      getValues: jest.fn().mockResolvedValue({ name: 'John Doe' }),
    };

    const mockPaymentMethodFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      tokenize: jest.fn().mockResolvedValue({ token: 'abc123' }),
    };

    // Mock component
    const mockComponent = new PaymentForm();

    // Cast to any to bypass type checker and directly set private properties
    (mockComponent as any).billingFormRef = mockBillingFormRef;
    (mockComponent as any).paymentMethodFormRef = mockPaymentMethodFormRef;

    // Mock event
    const mockEvent = { preventDefault: jest.fn() };

    // Mock event emitter
    const mockEmitter = { emit: jest.fn() };
    mockComponent.submitted = mockEmitter as any;

    // Call submit method
    await mockComponent.submit(mockEvent);

    // Expectations
    expect(mockEvent.preventDefault).toHaveBeenCalled();
    expect(mockBillingFormRef.validate).toHaveBeenCalled();
    expect(mockPaymentMethodFormRef.validate).toHaveBeenCalled();
    expect(mockBillingFormRef.getValues).toHaveBeenCalled();
    expect(mockPaymentMethodFormRef.tokenize).toHaveBeenCalledWith(mockComponent.clientId, { email: mockComponent.email, name: 'John Doe' }, mockComponent.accountId);
    expect(mockEmitter.emit).toHaveBeenCalledWith({ token: 'abc123' });
  });

  // Event Testing
  it('should fire the "submitted" event with correct data when the form is submitted', async () => {
    // Mock billingFormRef and paymentMethodFormRef
    const mockBillingFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      getValues: jest.fn().mockResolvedValue({
        name: 'John Doe',
        address_line1: 'Address 1',
        address_line2: 'Address 2',
        address_city: 'City',
        address_state: 'State',
        address_postal_code: 'Postal Code',
      }),
    };

    const mockPaymentMethodFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      tokenize: jest.fn().mockResolvedValue({ token: 'abc123' }),
    };

    // Mock component
    const mockComponent = new PaymentForm();
    mockComponent.email = 'john@example.com'; // set email prop

    // Cast to any to bypass type checker and directly set private properties
    (mockComponent as any).billingFormRef = mockBillingFormRef;
    (mockComponent as any).paymentMethodFormRef = mockPaymentMethodFormRef;

    // Mock event
    const mockEvent = { preventDefault: jest.fn() };

    // Mock event emitter
    const mockEmitter = { emit: jest.fn() };
    mockComponent.submitted = mockEmitter as any;

    // Call submit method
    await mockComponent.submit(mockEvent);

    // Define expected payload
    const expectedPayload = { token: 'abc123' };

    // Expectations
    expect(mockEmitter.emit).toHaveBeenCalledWith(expectedPayload);
  });

  // Render Testing
  it('should correctly render the child components based on the given props and state', async () => {
    // Render the component
    const { root } = await newSpecPage({
      components: [PaymentForm],
      template: () => (
        <justifi-payment-form
          bankAccount
          card
          email="test@test.com"
          clientId="abc123"
          accountId="def456"
          submit-button-text="Submit"
          payment-method-form-type="card"
        />),
    });

    // Assert that the rendered output is correct
    expect(root).toMatchSnapshot();
  });

  it('emits error event when submit fails', async () => {
    const clientId = 'some-client';

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [PaymentForm, PaymentMethodSelector, PaymentMethodForm, BillingForm],
      template: () => <justifi-payment-form clientId={clientId} onError-event={errorSpy} />,
    });

    page.rootInstance.paymentMethodFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      tokenize: jest.fn().mockRejectedValue(new Error('Tokenize error')),
    };;
    page.rootInstance.billingFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      getValues: jest.fn().mockResolvedValue({}),
    };

    await page.waitForChanges();

    const submitButton = page.root.shadowRoot.querySelector('[data-testid="submit-button"]') as HTMLButtonElement;

    submitButton.click();

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        errorCode: 'unknown-error',
        message: 'Tokenize error',
        severity: 'error',
      }
    }));
  });

  it('emits error event when tokenize fails', async () => {
    const clientId = 'some-client';

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [PaymentForm, PaymentMethodSelector, PaymentMethodForm, BillingForm],
      template: () => <justifi-payment-form clientId={clientId} onError-event={errorSpy} />,
    });

    page.rootInstance.paymentMethodFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      tokenize: jest.fn().mockResolvedValue({ error: { message: 'Tokenize error' } }),
    };

    page.rootInstance.billingFormRef = {
      validate: jest.fn().mockResolvedValue({ isValid: true }),
      getValues: jest.fn().mockResolvedValue({}),
    };

    await page.waitForChanges();

    const submitButton = page.root.shadowRoot.querySelector('[data-testid="submit-button"]') as HTMLButtonElement;

    submitButton.click();

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        errorCode: 'tokenize-error',
        message: 'Tokenize error',
        severity: 'error',
      }
    }));
  });
});
