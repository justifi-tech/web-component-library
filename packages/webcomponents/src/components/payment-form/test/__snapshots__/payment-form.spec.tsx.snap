// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-payment-form should correctly render the child components based on the given props and state 1`] = `
<justifi-payment-form payment-method-form-type="card" submit-button-text="Submit">
  <template shadowrootmode="open">
    <form class="gy-3 row">
      <div class="col-12">
        <justifi-payment-method-selector selectedpaymentmethodtype="card"></justifi-payment-method-selector>
      </div>
      <div class="col-12">
        <justifi-payment-method-form iframeorigin="https://js.justifi.ai/v2" payment-method-form-type="card"></justifi-payment-method-form>
      </div>
      <div class="col-12">
        <justifi-billing-form legend="Billing Info"></justifi-billing-form>
      </div>
      <slot name="insurance"></slot>
      <div class="col-12">
        <button class="btn btn-primary jfi-submit-button" data-testid="submit-button" type="submit">
          Submit
        </button>
      </div>
    </form>
  </template>
</justifi-payment-form>
`;
